### PR TITLE
Escape terminator characters in filter arguments

### PR DESCRIPTION
**Pull request by Depaulicious**
Self explanatory. I lost an hour trying to understand why it wouldn't trim videos correctly when I specified the timestamps as `HH:MM:SS.fraction`. Damnit.

### CODE REVIEW

File: ffmpeg/nodes.py [hunk 1]
```diff
@@ -51,8 +52,21 @@ class FilterNode(Node):
...
         params_text = self._name
-        arg_params = ['{}'.format(arg) for arg in self._args]
-        kwarg_params = ['{}={}'.format(k, self._kwargs[k]) for k in sorted(self._kwargs)]
+
+        # Helper function to escape uncomfortable characters
+        def escape_chars(seq, keys):
+            for k in keys:
+                if not isinstance(seq[k], basestring):
+                    continue
+                for ch in "[]=;:,":
```
Comment (kkroening) [inline ffmpeg/nodes.py]: Should also escape `\\` chars, e.g. `\\` => `\\\\`.

**Comment by kkroening** [comment]
Cool - I was wondering when we would run into this. It's probably going to need some additional tweaking as more edge cases are uncovered but it's a good start.

Gotta figure out why tests are failing though. I'll take a look

File: ffmpeg/nodes.py [hunk 2]
```diff
@@ -51,8 +52,21 @@ class FilterNode(Node):
...
+                    seq[k] = seq[k].replace(ch, "\\\\"+ch)
```
Comment (kkroening) [inline ffmpeg/nodes.py]: Should this be `"\\"+ch`?

File: ffmpeg/nodes.py [hunk 3]
```diff
@@ -51,8 +52,21 @@ class FilterNode(Node):
     """FilterNode"""
     def _get_filter(self):
         params_text = self._name
-        arg_params = ['{}'.format(arg) for arg in self._args]
-        kwarg_params = ['{}={}'.format(k, self._kwargs[k]) for k in sorted(self._kwargs)]
+
+        # Helper function to escape uncomfortable characters
+        def escape_chars(seq, keys):
```
Comment (kkroening) [inline ffmpeg/nodes.py]: If we pull this out to a global function we can put a test around it:
```
def _escape_chars(text):
    """Helper function to escape uncomfortable characters."""
    for ch in '\\[]=;:,':
        text = text.replace(ch, '\\' + ch)
    return text

# in test_ffmpeg.py:
def test_escape_chars():
    assert ffmpeg.nodes._escape_chars('a[b]c=d;e,\\f') == 'a\\[b\\]c\\=d\\;e\\,\\\\f'
```
Then `FilterNode._get_filter` becomes..
```
def _get_filter(self):
    params_text = self._name
    _args = [_escape_chars(str(x)) for x in self._args]
    _kwargs = {k: _escape_chars(str(v)) for k, v in self._kwargs.items()}
    ...
```

Comment (kkroening) [inline ffmpeg/nodes.py]: Wow, I ran a few scenarios against `ffmpeg` using `drawtext`, and `ffmpeg`'s string unescaping is super bizarre. Some chars only need a single backslash before them, some require two backslashes, and some require _eight backslashes_.

Here's an example that draws the string `a\b`:
```
ffmpeg -i ffmpeg/tests/sample_data/dummy.mp4 -filter_complex '[0]drawtext=x=0:y=100:text=a\\\\\\\\b:fontfile=/Library/Fonts/Arial.ttf[1]' -map '[1]' out.mp4
```
Yup, _eight backslashes_.

I found [this stackoverflow article](https://stackoverflow.com/questions/10725225/ffmpeg-single-quote-in-drawtext) talking about it:
> Special character escapes are like violence: if they're not solving your problem, you're not using enough.
> `ffmpeg -i test.mpg -vf drawtext=text="It\\\\\'s so easy"`

I'm gonna try to see if there's any method to the madness and whether there's a way to completely correctly escape. It's nice to have this done in a tool like `ffmpeg-python` so that hopefully nobody else has to deal with ffmpeg's insanity.

Comment (Depaulicious) [inline ffmpeg/nodes.py]: OMG this is crazy. In fact I did notice that I only needed one backslash when running it on the shell, and two with python, which resulted in 4 backslashes.

I'm currently running my code in my testing branch and I'm keeping it for the moment as it works for timestamps (which are basically the only thing that needs escaping in our project, though we could avoid it using `SS.fraction` only).

Let me know if I should commit the change you proposed or if you're going to do it some other way.

> The text is being parsed a couple times, so you not only have to escape the quote, you also have to escape the slash escaping the quote. Twice.

If what the StackOverflow article says is right, we should do a few real-world tests with ffmpeg, find out how many backslashes are required for a particular filter or character, then make an *intelligent escaper* out of these numbers. The escaper would escape characters in `trim` arguments with `r'\\'`, `text` argument of `drawtext` with `r'\\\`, etc., or fallback to some default.

**Comment by kkroening** [comment]
Alright so as mentioned above, there are a lot of weird ffmpeg string-escaping quirks. I tried a bunch of scenarios to try to figure out how ffmpeg unescapes strings and still don't fully understand it. Looks like there are multiple layers of escaping that need to happen. I think I'm close to figuring out the quirks but have to wrap up for now and try to get to it tomorrow.

File: ffmpeg/nodes.py (hunk 3 above)
Comment (kkroening) [inline ffmpeg/nodes.py]: Here are my science results:
```
`'`: 7 backslashes in drawtext text param, 3 in fontfile; 3 in param names; 1 in filter name
`\`: 7 backslashes in drawtext text param, 3 in fontfile; 3 in param names; 1 in filter name
`:`: 2 backslashes
`,`: 1 backslash
`[`: 1 backslash
`]`: 1 backslash
`=`: 1 backslash
```
I was thinking something like this:
```
args = [escape_chars(str(x, '=:')) for x in self._args]
kwargs = {}
for k, v in self._kwargs.items():
    k = escape_chars(k, '\\\'=:')
    v = escape_chars(str(v), '\\\'=:')
    kwargs[k] = v

arg_params = ['{}'.format(arg) for arg in args]
kwarg_params = ['{}={}'.format(k, kwargs[k]) for k in sorted(kwargs)]
params = ':'.join(arg_params + kwarg_params)

params_text = self._name
if params:
    params_text += '={}'.format(':'.join(params))
return escape_chars(params_text, '\\\'[],;')
```
And then another `escape_chars` for rendering the `text` param of `drawtext`. I think this produces similar results to what's above. But need more science

Comment (kkroening) [inline ffmpeg/nodes.py]: I wonder if there's any way to encode unicode chars in ffmpeg args. You'd think there would be, since people might want to render Chinese text over a video

**Comment by kkroening** [comment]
Ok I updated the string escaping to be pretty robust.

I added the `drawtext` filter since its `text` arg has slightly more complicated string escaping that other filter args.

The string escaping actually makes sense to me now. The reason why there are so many backslashes is that `\` and `'` characters are escape/unescaped numerous times as the command-line is constructed/destructed. It's actually intentional by the ffmpeg designers so that tools like `ffmpeg-python` can programmatically escape things in a well-defined manner. It's a pain to do by hand, but libraries like this can automate it and make it a non-issue.

The tests are table-driven to make sure all of the characters are escaped as desired, both in the `drawtext` `text` arg and normal args. This is nearly a direct translation of the science table I posted earlier with the character -> expected backslash count:
```
    expected_backslash_counts = {
        'x': 0,
        '\'': 7,
        '\\': 7,
        '%': 4,
        ':': 2,
        ',': 1,
        '[': 1,
        ']': 1,
        '=': 2,
        '\n': 0,
    }
    for ch, expected_backslash_count in expected_backslash_counts.items():
        expected = '{}{}'.format('\\' * expected_backslash_count, ch)
        actual = _get_drawtext_text_repr(ch)
        assert expected == actual
```
So basically with these tests, we should be pretty confident that we escape all things correctly in all circumstances. And if there are any extra edge cases, we can just add it to the above test table and then see the tests pass once we fix the code.

**Comment by kkroening** [comment]
Here it is in action:
```
(ffmpeg
    .input('in.mp4')
    .drawtext(
        u'some quotes: \'yay!\'\na newline\nand more stuff :,\\\'\nthe end',
        x=0,
        y=100,
        fontfile='LiberationSans.ttf', box=1
    )
    .output('out.mp4')
    .overwrite_output()
    .run()
)
```
Args:
```
['-i',
 'in.mp4',
 '-filter_complex',
 "[0]drawtext=box=1:fontfile=LiberationSans.ttf:text=some quotes\\\\: \\\\\\\\\\\\\\'yay!\\\\\\\\\\\\\\'\na newline\nand more stuff \\\\:\\,\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\'\nthe end:y=100[v0]",
 '-map',
 '[v0]',
 'out.mp4']
```
Output:
out0001

**Comment by kkroening** [comment]
I'm gonna merge this so I can pull it into the other branches. Let me know if you find any issues with it.
